src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -85,36 +130,75 @@ public static byte[] Sign(byte[] content!!, AsymmetricAlgorithm key!!, HashAlgor
+            Span<byte> toBeSignedBuffer = buffer;
+            byte[]? rentedToBeSignedBuffer = null;
+            int signatureBytesWritten;
+
+            // It is possible for toBeSigned to be bigger than the COSE message length that we used to determine the size of our buffer.
+            // we rent a bigger buffer if that's the case.
+            if (buffer.Length < expectedToBeSignedSize)
+            {
+                rentedToBeSignedBuffer = ArrayPool<byte>.Shared.Rent(expectedToBeSignedSize);
+                toBeSignedBuffer = rentedToBeSignedBuffer;
+            }
+
+            try
+            {
+                int toBeSignedBytesWritten = CreateToBeSigned(SigStructureCoxtextSign1, encodedProtectedHeaders, content, toBeSignedBuffer);
+                ReadOnlySpan<byte> encodedToBeSigned = buffer.Slice(0, toBeSignedBytesWritten);
+
+                if (keyType == KeyType.ECDsa)
+                {
+                    signatureBytesWritten = SignWithECDsa((ECDsa)key, encodedToBeSigned, hashAlgorithm, buffer);
+                }
+                else
+                {
+                    signatureBytesWritten = SignWithRSA((RSA)key, encodedToBeSigned, hashAlgorithm, buffer);
+                }
+            }
+            finally
+            {
+                if (rentedToBeSignedBuffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(rentedToBeSignedBuffer);
+                }
+            }
+
+            writer.WriteByteString(buffer.Slice(0, signatureBytesWritten));
+
             writer.WriteEndArray();
 
-            return writer.Encode();
+            return writer.Encode(buffer);
         }
 
         [UnsupportedOSPlatform("browser")]
-        private static byte[] SignWithECDsa(ECDsa key, byte[] data, HashAlgorithmName hashAlgorithm)
-            => key.SignData(data, hashAlgorithm);
-
-        [UnsupportedOSPlatform("browser")]
-        private static byte[] SignWithRSA(RSA key, byte[] data, HashAlgorithmName hashAlgorithm)
-            => key.SignData(data, hashAlgorithm, RSASignaturePadding.Pss);
-
-        internal static byte[] CreateEncodedProtectedHeader(KeyType algType, HashAlgorithmName hashAlgorithm)
+        private static int SignWithECDsa(ECDsa key, ReadOnlySpan<byte> data, HashAlgorithmName hashAlgorithm, Span<byte> destination)
         {
-            var writer = new CborWriter();
-            writer.WriteStartMap(1);
-            writer.WriteInt32(KnownHeaders.Alg);
-            writer.WriteInt32(GetCoseAlgorithmHeaderFromKeyTypeAndHashAlgorithm(algType, hashAlgorithm));
-            writer.WriteEndMap();
-            return writer.Encode();
+#if NETSTANDARD2_0 || NETFRAMEWORK
+            byte[] signature = key.SignData(data.ToArray(), hashAlgorithm);
+            signature.CopyTo(destination);
+            return signature.Length;
+#else
+            bool success = key.TrySignData(data, destination, hashAlgorithm, out int bytesWritten);
+            Debug.Assert(success);
```

Comment:
```suggestion
            if (!success)
            {
                Debug.Fail("TrySignData failed with a pre-calculated destination");
                throw new CryptographicException();
            }
```

While the asymmetric algorithm types aren't multi-thread safe, and the COSE types aren't multi-thread safe, we shouldn't let someone doing something silly like regenerating the key to a bigger size on a different thread cause us to run down weird code paths.

Reply:
Got it, but if we throw, we don't need Debug.Fail, right? (it would be redundant).

Reply:
The Debug.Fail provides documentation that it's an unexpected code path, largely justifying the exception with no specific message.  And gives a very specific place to look if something goes wrong in a debug build :)